tests/python/frontend/tflite/test_forward.py

```diff
@@ -161,6 +163,39 @@ def with_fused_activation_function(input_tensor, fn_name):
         return math_ops.tanh(input_tensor)
     raise AssertionError("Unknown fused_activation_function {}".format(fn_name))
 
+def _test_split(in_shape, axis, num_Splits, dtype):
+    '''internal split tester taking as parameters in_shape, number of tensors to split into
+       and dtype (data type)'''
+    np_data = np.random.uniform(-5, 5, size=in_shape).astype(dtype)
+    with tf.Graph().as_default():
+        in_data = array_ops.placeholder(shape=in_shape, dtype=dtype)
+        out = array_ops.split(in_data, num_Splits, axis=axis)
+        compare_tflite_with_tvm(np_data, 'Placeholder:0',  [in_data], out,
+                                out_names=[f'out:{n}' for n in range(num_Splits)])
```

Comment:
Consider becomes  `out_names=[f'out_{n}:0' for n in range(num_Splits)])`. Maybe could help to pass CI.

Reply:
Thanks for the hint, but could you elaborate why the change is needed ?

Reply:
I couldn't make sure completely. I just think of when I use `tflite_convert` tool, tflite doesn't support `input:0`, just support `input`. So maybe in TFLite, `input:0 / input:1` will be handled as the same `input`. You could try.

Reply:
The backtrace looks like this. 

  File "/workspace/tests/python/frontend/tflite/test_forward.py", line 174, in _test_split
    out_names=[f'out:{n}' for n in range(num_Splits)])
  File "/workspace/tests/python/frontend/tflite/test_forward.py", line 148, in compare_tflite_with_tvm
    num_output=len(out_names), out_names=out_names)
  File "/workspace/tests/python/frontend/tflite/test_forward.py", line 72, in run_tvm_graph
    dtype_dict=dtype_dict)
  File "/workspace/python/tvm/relay/frontend/tflite.py", line 941, in from_tflite
    op_converter.convert_op_to_relay()
  File "/workspace/python/tvm/relay/frontend/tflite.py", line 113, in convert_op_to_relay
    self.exp_tab.set_expr(get_tensor_name(self.subgraph, tensor_idx), ret)
  File "/workspace/python/tvm/relay/frontend/common.py", line 273, in set_expr
    assert isinstance(expr, _expr.Expr)

That indicates an assertion failure at isinstance(expr, _expr.Expr) in set_expr in common.py that comes as a result of the set_expr call from convert_op_to_relay.

Tracing op.split -> _make.split shows that the expression received back should really be TupleWrapper(expr) . 

Finally getting this to fail on my machine at home , I see that expr that comes back is TupleWrapper(CallNode(Op(split), [Var(Placeholder, ty=TensorType([3], float32))], relay.attrs.SplitAttrs(0x36aadf0), []), 1)

Any further hints / thoughts ?

Reply:
Have you applied my suggestion? I find the backtrace is the same code of this PR.

Reply:
Thanks for the reminder. On the machine where I see a failure I see the following output with the changes you suggested. I need to work out where the 2 environments differ and that needs me to find some dedicated debug time for this one.

-        compare_tflite_with_tvm(np_data, 'Placeholder:0',  [in_data], out,
-                                out_names=[f'out:{n}' for n in range(num_Splits)])
+        compare_tflite_with_tvm([np_data], ['Placeholder:0'],  [in_data], out,
+                                out_names=[f'out_{n}:0' for n in range(num_Splits)])


$>
File "test_forward.py", line 786, in <module>
    test_forward_split()
  File "test_forward.py", line 179, in test_forward_split
    _test_split((3,), 0, 1, 'float32')
  File "test_forward.py", line 174, in _test_split
    out_names=[f'out_{n}:0' for n in range(num_Splits)])
  File "test_forward.py", line 148, in compare_tflite_with_tvm
    num_output=len(out_names), out_names=out_names)
  File "test_forward.py", line 72, in run_tvm_graph
    dtype_dict=dtype_dict)
  File "/home/ramana/tvm-test/tvm/python/tvm/relay/frontend/tflite.py", line 941, in from_tflite
    op_converter.convert_op_to_relay()
  File "/home/ramana/tvm-test/tvm/python/tvm/relay/frontend/tflite.py", line 113, in convert_op_to_relay
    self.exp_tab.set_expr(get_tensor_name(self.subgraph, tensor_idx), ret)
  File "/home/ramana/tvm-test/tvm/python/tvm/relay/frontend/common.py", line 273, in set_expr
    assert isinstance(expr, _expr.Expr)

Reply:
What version of tensorflow do we use inside the jenkins jobs ? Aha, it looks like the failure occurs if I use tensorflow 1.12 and not tensorflow 1.14 ! What's the minimum version of tf that we support ?

Switching to tf1.14 makes it just work (TM).

It'll be Friday before I can pick this up next.

Reply:
Our tf of jenkins should be 1.13.1

Reply:
I think that should be pinned along with other frameworks in our install scripts. We should take that discussion up in https://discuss.tvm.ai/t/reproducibility-of-ci-environment/1537. I'll continue to debug this as I'm seeing differences in relay output for the same input depending on whether I use tf 1.13 or tf1.14

Reply:
I managed to steal some time today. Here's the difference in output for tf 1.14 vs tf 1.13 

The tflite input_details and output_details from 
```

def run_tflite_graph(tflite_model_buf, input_data):
    """ Generic function to execute TFLite """
    input_data = convert_to_list(input_data)

    interpreter = interpreter_wrapper.Interpreter(model_content=tflite_model_buf)
    interpreter.allocate_tensors()

    input_details = interpreter.get_input_details()
    output_details = interpreter.get_output_details()
```

This is the output for tf 1.13.1
```
2019-07-18 14:48:56.601456: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
tflite input_details are, [{'name': 'Placeholder', 'index': 0, 'shape': array([2, 6], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}]
tflite output_Details are, [{'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}, {'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}, {'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}, {'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}, {'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}, {'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}]

```

And this is the output for tf 1.14 

`tflite input_details are, [{'name': 'Placeholder', 'index': 0, 'shape': array([2, 6], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}]
tflite output_Details are, [{'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}, {'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}, {'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}, {'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}, {'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}, {'name': 'split', 'index': 1, 'shape': array([2, 1], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}]
tflite_output is [array([[2.0315871],
`
Which to me sounds like there is a bug between tf1.13.x and tf1.14.x

Any ideas ?

Reply:
Could you help to list the difference? Because the print log information has too much. I can not handle the key point.

Reply:
@FrozenGene , I've highlighted in bold (1.13 ) and bold+italic (1.14)

Reply:
Thanks a lot for your help fixing up the other bits.

For the input testcase below

` _test_split((3,), 0, 3, 'float32')`

The relay output difference between 1.13 and 1.14 is as below. 

For 1.13 
v0.0.3
def @main(%Placeholder: Tensor[(3,), float32]) -> (Tensor[(1,), float32], Tensor[(1,), float32], Tensor[(1,), float32]) 
{
  %0 = split(%Placeholder, indices_or_sections=int64(3)) /* ty=(Tensor[(1,), float32], Tensor[(1,), float32], Tensor[(1,), float32]) */;
  **%1 = %0.0;
  (%1, %1, %1)**
}


For 1.14


`v0.0.3
  def @main(%Placeholder: Tensor[(3,), float32]) -> (Tensor[(1,), float32], Tensor[(1,), float32], Tensor[(1,), float32]) {
    %0 = split(%Placeholder, indices_or_sections=int64(3)) /* ty=(Tensor[(1,), float32], Tensor[(1,), float32], Tensor[(1,), float32]) */;
    **_%1 = %0.0;
    %2 = %0.1;
    %3 = %0.2;
    (%1, %2, %3)_**
  }
 
And this I think is a manifestation of the difference we see in the outputs of the tensorflow lite interpreter as I've said above. 

From my understanding the output we get for 1.14 is what is expected of the split operator, the output from 1.13 is just wrong as it only takes the first sub-tensor and broadcasts it across the rest. 

I haven't yet managed to find anything that says tflite's interpretation of split has changed between 1.13 and 1.14.